internal/venice-common/src/main/java/com/linkedin/venice/writer/VeniceWriter.java

```diff
@@ -1760,30 +1717,68 @@ public void sendControlMessage(
              * Not a super clean way to match the exception, but unfortunately, since it is wrapped inside of an
              * {@link ExecutionException}, there may be no other way.
              */
-            String errorMessage =
-                "Caught a UNKNOWN_TOPIC_OR_PARTITION error, attempt " + attempt + "/" + maxAttemptsWhenTopicMissing;
             if (attempt < maxAttemptsWhenTopicMissing) {
               attempt++;
               updateCheckSum = false; // checksum has already been updated, and should not be updated again for retries
-              logger.warn("{}, will sleep {} ms before the next attempt.", errorMessage, sleepTimeMsWhenTopicMissing);
+              logger.warn(
+                  "Caught a UNKNOWN_TOPIC_OR_PARTITION error, attempt {}/{}",
+                  attempt,
+                  maxAttemptsWhenTopicMissing);
             } else {
-              throw new VeniceException(errorMessage + ", will bubble up.");
+              throw new VeniceException(
+                  "Caught a UNKNOWN_TOPIC_OR_PARTITION error, attempt " + attempt + "/" + maxAttemptsWhenTopicMissing
+                      + ", will bubble up.");
             }
-          } else if (ExceptionUtils.recursiveClassEquals(e, PubSubTopicAuthorizationException.class)) {
-            throw new VeniceResourceAccessException(
-                "You do not have permission to write to this store. Please check that ACLs are set correctly.",
-                e);
           } else {
-            throw new VeniceException(
-                "Got an exception while trying to send a control message ("
-                    + ControlMessageType.valueOf(controlMessage).name() + ")",
-                e);
+            handleControlMessageProducingException(e, controlMessageType);
           }
         }
       }
     }
   }
 
+  private void handleControlMessageProducingException(Exception e, ControlMessageType controlMessageType) {
+    if (ExceptionUtils.recursiveClassEquals(e, PubSubTopicAuthorizationException.class)) {
+      throw new VeniceResourceAccessException(
+          "You do not have permission to write to this store. Please check that ACLs are set correctly.",
+          e);
+    } else {
+      throw new VeniceException(
+          "Got an exception while trying to send a control message (" + controlMessageType.name() + ")",
+          e);
+    }
+  }
+
+  /**
+   * Main function for sending control messages. Synchronization is most minimal and there is no error checking.
+   *
+   * See also:
+   * {@link #sendControlMessageWithRetriesForNonExistentTopic(ControlMessage, int, Map, PubSubProducerCallback, LeaderMetadataWrapper)}
+   */
+  public CompletableFuture<PubSubProduceResult> sendControlMessage(
```

Comment:
I think we need to ensure that the callers (e.g. broadcastControlMessage) wait on this future to keep the old behavior.